quickwit/quickwit-indexing/src/actors/doc_processor.rs

```diff
@@ -151,25 +262,43 @@ impl DocProcessor {
         source_id: String,
         doc_mapper: Arc<dyn DocMapper>,
         indexer_mailbox: Mailbox<Indexer>,
-    ) -> Self {
+        vrl_settings: Option<VrlSettings>,
+    ) -> anyhow::Result<Self> {
         let schema = doc_mapper.schema();
         let timestamp_field_opt = doc_mapper.timestamp_field(&schema);
-        Self {
+        let vrl_program = match vrl_settings {
+            Some(settings) => Some(VrlProgram::try_from(settings)?),
+            None => None,
+        };
+
+        let doc_processor = Self {
             doc_mapper,
             indexer_mailbox,
             timestamp_field_opt,
             counters: DocProcessorCounters::new(index_id, source_id),
             publish_lock: PublishLock::default(),
-        }
+            transform_layer: vrl_program,
+        };
+
+        Ok(doc_processor)
     }
 
     fn prepare_document(
-        &self,
+        &mut self,
         doc_json: String,
         ctx: &ActorContext<Self>,
     ) -> Result<PreparedDoc, PrepareDocumentError> {
-        // Parse the document
         let _protect_guard = ctx.protect_zone();
+        self._prepare_document(doc_json)
+    }
+
+    pub fn _prepare_document(&mut self, doc_json: String) -> Result<PreparedDoc, PrepareDocumentError> {
```

Comment:
Oh, I see what you did here. Come on, you know this is ugly. Please, propose a better alternative.